redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt

```diff
@@ -62,64 +54,141 @@ internal class ViewLazyColumn<A : AppService>(
-          gravity = Gravity.CENTER_HORIZONTAL
+    override fun onViewRecycled(holder: ViewHolder<A>) {
+      holder.ziplineScope.close()
+    }
+
+    private class ViewHolder<A : AppService>(
+      val container: FrameLayout,
+      treehouseApp: TreehouseApp<A>,
+      widgetSystem: TreehouseView.WidgetSystem<A>,
+    ) : RecyclerView.ViewHolder(container) {
+
+      val ziplineScope = ZiplineScope()
+
+      var interval: LazyListIntervalContent? = null
+
+      val blankView = FrameLayout(itemView.context).apply {
+        layoutParams = FrameLayout.LayoutParams(
+          MATCH_PARENT,
+          TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 1000F, itemView.resources.displayMetrics).toInt(),
+        )
+      }
+
+      var placeholderTreehouseWidgetViewLoadState = LoadState.Loading
+      val placeholderTreehouseWidgetViewCodeListener = object : TreehouseView.CodeListener() {
+        override fun onCodeLoaded(initial: Boolean) {
+          val newState = LoadState.Loaded
+          if (placeholderTreehouseWidgetViewLoadState == newState) return
+          placeholderTreehouseWidgetViewLoadState = newState
+          onLoadStateChange()
+        }
+      }
+      val placeholderTreehouseWidgetView: TreehouseWidgetView<A> = TreehouseWidgetView(container.context, widgetSystem, placeholderTreehouseWidgetViewCodeListener)
+        .apply {
+          treehouseApp.renderTo(this)
+
+          layoutParams = FrameLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT).apply {
+            gravity = Gravity.CENTER_HORIZONTAL
+          }
         }
 
-        container.addView(this)
+      var itemTreehouseWidgetViewLoadState = LoadState.Loading
+      val itemTreehouseWidgetViewCodeListener = object : TreehouseView.CodeListener() {
+        override fun onCodeLoaded(initial: Boolean) {
+          val newState = LoadState.Loaded
+          if (itemTreehouseWidgetViewLoadState == newState) return
+          itemTreehouseWidgetViewLoadState = newState
+          onLoadStateChange()
+        }
       }
+      val itemTreehouseWidgetView: TreehouseWidgetView<A> = TreehouseWidgetView(container.context, widgetSystem, itemTreehouseWidgetViewCodeListener)
+        .apply {
+          treehouseApp.renderTo(this)
+
+          layoutParams = FrameLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT).apply {
+            gravity = Gravity.CENTER_HORIZONTAL
+          }
+        }
+
+      fun onLoadStateChange() {
+        val interval = checkNotNull(interval)
+
+        val holder: ViewHolder<A> = this
+        holder.container.removeAllViews()
+        when {
+          holder.itemTreehouseWidgetViewLoadState == LoadState.Loaded -> {
+            // Show the item
+            holder.itemTreehouseWidgetView.isVisible = true
+            holder.container.addView(holder.itemTreehouseWidgetView)
+          }
+          holder.placeholderTreehouseWidgetViewLoadState == LoadState.Loading -> {
```

Comment:
Reorder the 2nd and 3rd when clauses? That way they’re processed in order of preference. (item, placeholder, blank)